src/snowflake/cli/_plugins/workspace/manager.py

```diff
@@ -31,13 +33,24 @@ def __init__(self, project_definition: ProjectDefinition, project_root: Path):
         self._entities_cache: Dict[str, Entity] = {}
         self._project_definition: DefinitionV20 = project_definition
         self._project_root = project_root
-        self._default_role = default_role()
-        if self._default_role is None:
-            self._default_role = get_sql_executor().current_role()
-        self.default_warehouse = None
-        cli_context = get_cli_context()
-        if cli_context.connection.warehouse:
-            self.default_warehouse = to_identifier(cli_context.connection.warehouse)
+        self._default_role = self._UNSET
+        self._default_warehouse = self._UNSET
+
+    def _get_default_role(self):
+        if self._default_role is self._UNSET:
+            self._default_role = default_role()
+            if self._default_role is None:
+                self._default_role = get_sql_executor().current_role()
+        return self._default_role
+
+    def _get_default_warehouse(self):
+        if self._default_warehouse is self._UNSET:
+            cli_context = get_cli_context()
+            if cli_context.connection.warehouse:
+                self._default_warehouse = to_identifier(
+                    cli_context.connection.warehouse
+                )
+        return self._default_warehouse
```

Comment:
there is no connection warehouse, isn't this going to return UNSET? That doesn't sound right. The role logic doesn't have this issue because if assigns `None` even if the fallback logic fails to produce a valid role.

Reply:
@sfc-gh-bdufour good catch, I updated the implementation to use `@cached_property` in the `ActionContext` since we don't need to be caching them in `WorkspaceManager`